### PR TITLE
bevy_reflect: Improve serialization format

### PR DESCRIPTION
**Pull request by MrGVSV**
> Note: See #5723 for an even more improved (but slightly more controversial) format

# Objective

The current serialization format used by `bevy_reflect` is both verbose and error-prone. Taking the following structs[^1] for example:

```rust
// -- src/inventory.rs

#[derive(Reflect)]
struct Inventory {
  id: String,
  max_storage: usize,
  items: Vec<Item>
}

#[derive(Reflect)]
struct Item {
  name: String
}
```

Given an inventory of a single item, this would serialize to something like:

```rust
// -- assets/inventory.ron

{
  "type": "my_game::inventory::Inventory",
  "struct": {
    "id": {
      "type": "alloc::string::String",
      "value": "inv001",
    },
    "max_storage": {
      "type": "usize",
      "value": 10
    },
    "items": {
      "type": "alloc::vec::Vec<alloc::string::String>",
      "list": [
        {
          "type": "my_game::inventory::Item",
          "struct": {
            "name": {
              "type": "alloc::string::String",
              "value": "Pickaxe"
            },
          },
        },
      ],
    },
  },
}
```

Aside from being really long and difficult to read, it also has a few "gotchas" that users need to be aware of if they want to edit the file manually. A major one is the requirement that you use the proper keys for a given type. For structs, you need `"struct"`. For lists, `"list"`. For tuple structs, `"tuple_struct"`. And so on.

It also ***requires*** that the `"type"` entry come before the actual data. Despite being a map— which in programming is almost always orderless by default— the entries need to be in a particular order. Failure to follow the ordering convention results in a failure to deserialize the data.

This makes it very prone to errors and annoyances.

## Solution

Using #4042, we can remove a lot of the boilerplate and metadata needed by this older system. Since we now have static access to type information, we can simplify our serialized data to look like:

```rust
// -- assets/inventory.ron

{
  "my_game::inventory::Inventory": {
    "id": "inv001",
    "max_storage": 10,
    "items": [
      {
        "name": "Pickaxe"
      },
    ],
  },
},
```

This is much more digestible and a lot less error-prone (no more key requirements and no more extra type names).

#### Custom Serialization

Additionally, this PR adds the opt-in ability to specify a custom serde implementation to be used rather than the one created via reflection. For example[^1]:

```rust
// -- src/inventory.rs

#[derive(Reflect, Serialize)]
#[reflect(Serialize)]
struct Item {
  #[serde(alias = "id")]
  name: String
}
```

```rust
// -- assets/inventory.ron

{
  "my_game::inventory::Inventory": {
    "id": "inv001",
    "max_storage": 10,
    "items": [
      (
        id: "Pickaxe"
      )
    ],
  },
},
```

Because we specified a custom `Serialize` implementation with `#[reflect(Serialize)]`, we actually get the RON-style syntax for a struct.

> Note: If we had done this for `Inventory`, it would still define the `type` and continue to serialize the data normally.
> <details>
> <summary><strong>See Example</strong></summary>
>
> ```rust
> // -- assets/inventory.ron
> 
> {
> "my_game::inventory::Inventory": (
>  id: "inv001",
>  max_storage: 10,
>  items: [
>    (
>      name: "Pickaxe"
>    ),
>  ],
> ),
> }
> ```
>
> </details>

By allowing users to define their own serialization methods, we do two things:

1. We give more control over how data is serialized/deserialized to the end user
2. We avoid having to re-define serde's attributes and forcing users to apply both (e.g. we don't need a `#[reflect(alias)]` attribute).

#### Tuples!

Tuples and tuple-likes now use serde's `de/serialize_tuple` method. This provides better semantics— being fixed-size sequences, we should be able to indicate that size at when de/serializing. It also addresses some of serde's own limitations (see sections below).

###### Tuples

Obviously, tuples now use this format. This is better because tuples do have a maximum size according to [serde](https://docs.rs/serde/latest/serde/trait.Serialize.html#impl-Serialize-for-(T0%2C%20T1%2C%20T2%2C%20T3%2C%20T4%2C%20T5%2C%20T6%2C%20T7%2C%20T8%2C%20T9%2C%20T10%2C%20T11%2C%20T12%2C%20T13%2C%20T14%2C%20T15)). I don't think we actually run into this limitation, but it's a good reminder to avoid large tuples in case we want to be compatible with other formats.

```js
{
  "(f32, f32)": (1.0, 1.0)
}
```

<details>
<summary>Old Format</summary>

```js
{
  "type": "(f32, f32)",
  "tuple": [
    {
      "type": "f32",
      "value": 1.0
    },
    {
      "type": "f32",
      "value": 1.0
    }
  ]
}
```

</details>

###### Tuple Structs

This also applies to tuple structs. Though, it should be noted that there _is_ a dedicated `de/serialize_tuple_struct` method in serde. However, it requires a static name, which is not yet provided by `TypeInfo`. Even if it was, we have no way of handling `Dynamic***` types (see https://github.com/bevyengine/rfcs/pull/56).

```js
{
  "my_crate::Bar": ("Hello World!")
}
```

<details>
<summary>Old Format</summary>

```js
{
  "type": "my_crate::Bar",
  "tuple_struct": [
    {
      "type": "alloc::string::String",
      "value": "Hello World!"
    }
  ]
}
```

</details>

###### Arrays

It may be a bit surprising to some, but arrays now also use the tuple format. This is because they essentially _are_ tuples (a sequence of values with a fixed size), but only allow for homogenous types. Additionally, this is how RON handles them and is probably a result of the 32-capacity limit imposed on them (both by [serde](https://docs.rs/serde/latest/serde/trait.Serialize.html#impl-Serialize-for-%5BT%3B%2032%5D) and by [bevy_reflect](https://docs.rs/bevy/latest/bevy/reflect/trait.GetTypeRegistration.html#impl-GetTypeRegistration-for-%5BT%3B%2032%5D)).

```js
{
  "[i32; 3]": (1, 2, 3)
}
```

<details>
<summary>Old Format</summary>

```js
{
  "type": "[i32; 3]",
  "array": [
    {
      "type": "i32",
      "value": 1
    },
    {
      "type": "i32",
      "value": 2
    },
    {
      "type": "i32",
      "value": 3
    }
  ]
}
```

</details>

---

## Changelog

* Re-worked serialization/deserialization for reflected types
* Added `TypedReflectDeserializer` for deserializing data with known `TypeInfo`
* Renamed `ReflectDeserializer` to `UntypedReflectDeserializer` 
* ~~Replaced usages of `deserialize_any` with `deserialize_map` for non-self-describing formats~~ Reverted this change since there are still some issues that need to be sorted out (in a separate PR). By reverting this, crates like `bincode` can throw an error when attempting to deserialize non-self-describing formats (`bincode` results in `DeserializeAnyNotSupported`)
* Tuples, tuple structs, and arrays are now all de/serialized as tuples (similar to RON)

## Migration Guide

* This PR reduces the verbosity of the scene format. Scenes will need to be updated accordingly:

```js
// Old format
{
  "type": "my_game::item::Item",
  "struct": {
    "id": {
      "type": "alloc::string::String",
      "value": "bevycraft:stone",
    },
    "tags": {
      "type": "alloc::vec::Vec<alloc::string::String>",
      "list": [
        {
          "type": "alloc::string::String",
          "value": "material"
        },
      ],
    },
}

// New format
{
  "my_game::item::Item": {
    "id": "bevycraft:stone",
    "tags": ["material"]
  }
}
```

* There are also some smaller changes to how certain types are serialized/deserialized. Tuples, tuple structs, and arrays are all de/serialized as standard tuples now (if using a de/serializer like RON):

```js
// Old format
{
  "type": "my_crate::Foo",
  "tuple": {
    "type": "(f32, f32)",
    "tuple": [
        {
          "type": "f32",
          "value": 1.0
        },
        {
          "type": "f32",
          "value": 1.0
        }
      ]
    },
    "tuple_struct": {
      "type": "my_crate::Bar",
      "tuple_struct": [
        {
          "type": "alloc::string::String",
          "value": "Hello World!"
        }
      ]
    },
    "array": {
      "type": "[i32; 3]",
      "array": [
        {
          "type": "i32",
          "value": 1
        },
        {
          "type": "i32",
          "value": 2
        },
        {
          "type": "i32",
          "value": 3
        }
      ]
    }
  }
}

// New format
{
  "my_crate::Foo": {
    "tuple": (1.0, 1.0),
    "tuple_struct": ("Hello World!"),
    "array": (1, 2, 3),
  }
}
```

> Entering fewer/more than the exact amount of values results in an error when deserializing.

* `ReflectDeserializer` has been renamed to `UntypedReflectDeserializer`

[^1]: Some derives omitted for brevity.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Previously discussed in #92 and #4153. To make life easier for reviewers, how does this compare to this proposals?

**Comment by MrGVSV** [comment]
> Previously discussed in #92 and #4153. To make life easier for reviewers, how does this compare to this proposals?

I actually hadn't seen #92, so thanks for pointing that out! 

In regards to #4153, though, I think this is a good intermediate step towards a syntax like that. The main "issue" with the syntax proposed in that PR is that it's pretty reliant on the RON format and likely wouldn't translate to JSON or other formats well— especially if they're generated by a non-Rust tool (see @james7132's [comment](https://github.com/bevyengine/bevy/issues/4153#issuecomment-1062134228)).

So I think this is (hopefully) a less controversial improvement to the format while we decide and finalize on something like #4153.

**Comment by PROMETHIA-27** [comment]
Anecdotally, this will be really useful in a surprising way; I'm using reflection to power messages for my editor, so cutting down on the verbosity is a performance win as well as a reliability win for how much types can change or move before messages become invalid between versions.

**Comment by alice-i-cecile** [comment]
> as well as a reliability win for how much types can change or move before messages become invalid between versions.

This would be appreciated. I've had cases where I trivially reorganized code and completely broke all of my serialized structs in frustrating ways that are hard for beginners to debug.

**Review by PROMETHIA-27 — APPROVED** [review]
LGTM!

File: crates/bevy_reflect/src/serde/de.rs [hunk 1]
```diff
@@ -29,260 +47,273 @@ impl<'a, 'de> DeserializeSeed<'de> for ReflectDeserializer<'a> {
...
+        }
 
-    fn visit_f32<E>(self, v: f32) -> Result<Self::Value, E>
-    where
-        E: de::Error,
-    {
-        Ok(Box::new(v))
+        match self.type_info {
+            TypeInfo::Struct(struct_info) => {
+                let mut dynamic_struct = deserializer.deserialize_map(StructVisitor {
```
Comment (nicopap) [inline crates/bevy_reflect/src/serde/de.rs:145]: Regarding `TypedReflectDeserializer`. I think it is ideal to use the specialized `deserialize_{tuple,tuple_stuct,etc.}` when matching `TypeInfo::{Tuple,TupleStruct,etc.}` because of format-specific optimizations. Using `deserlaize_seq` and `deserialize_map` in those circumstances will force the serialization format to serialize the plain text value of the field names etc. This might sound alright or unimportant for a plain-text format like `json` or `ron` (and even `msgpack`) but for optimized binary formats, the distinction is important, since it helps them:
- Elide field count info from the serialized format
- Elide field names as string as well

I'm writing my own deserializer. I'm going to try to do that and tell you how it looks like. Seems possible now that most strings in `TypeInfo` have `'static` lifetimes.

Comment (nicopap) [reply crates/bevy_reflect/src/serde/de.rs:145]: Oooof, Why the hell does `deserialize_struct` take a `&'static [&'static str]`? Only `&[&'static str]` would be a bazilion times better :(

Comment (MrGVSV) [reply crates/bevy_reflect/src/serde/de.rs:145]: Yep this is definitely the direction I think a lot of people would like to move towards. And yeah the whole `'static` requirement on everything makes it a bit trickier.

For now though, I'd consider such changes outside the scope of this PR. My goal here is to do something that's relatively uncontroversial (i.e. just reduce the number of "type" fields) so it can be approved/merged sooner.

Although I did enable custom serde formats (see description) as a kind of compromise towards that end goal in the meantime.

File: crates/bevy_reflect/src/serde/de.rs [hunk 2]
```diff
@@ -380,130 +397,194 @@ impl<'a, 'de> Visitor<'de> for MapVisitor<'a> {
...
-    where
-        D: serde::Deserializer<'de>,
-    {
-        deserializer.deserialize_seq(TupleVisitor {
-            registry: self.registry,
-        })
-    }
-}
-
 struct TupleVisitor<'a> {
```
Comment (nicopap) [inline crates/bevy_reflect/src/serde/de.rs:400]: Extreme tinny nit: I would move `TupleVisitor` impl just after the `TupleStructVisitor` impl, so that it's obvious the code for the two is very similar, and they are easier to identify.

**Review by nicopap — APPROVED** [review]
Nice work, awesome change! Just a tinny nit and a question. I'm a bit wary from the repetition, but I know from experience it's basically impossible to work around it.

File: crates/bevy_reflect/src/serde/de.rs [hunk 3]
```diff
@@ -12,6 +15,21 @@ pub trait DeserializeValue {
...
+/// The serialized data must take the form of a map containing the following entries:
+/// 1. `type`: The _full_ [type name]
+/// 2. `value`: The serialized value of the reflected type
+///
+/// > Note: The ordering is important here. `type` _must_ come before `value`.
+///
+/// [`DynamicStruct`]: crate::DynamicStruct
+/// [`DynamicList`]: crate::DynamicList
+/// [type name]: std::any::type_name
 pub struct ReflectDeserializer<'a> {
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_reflect/src/serde/de.rs:33]: No strong feelings on the rename; I can see the arguments on both sides.

File: crates/bevy_reflect/src/serde/de.rs [hunk 4]
```diff
@@ -29,260 +47,273 @@ impl<'a, 'de> DeserializeSeed<'de> for ReflectDeserializer<'a> {
...
     }
 }
 
-struct ReflectVisitor<'a> {
+struct ReflectDeserializerVisitor<'a> {
     registry: &'a TypeRegistry,
 }
 
-impl<'a, 'de> Visitor<'de> for ReflectVisitor<'a> {
+impl<'a, 'de> Visitor<'de> for ReflectDeserializerVisitor<'a> {
```
[the remark this answers is not in the document]
Comment (MrGVSV) [reply crates/bevy_reflect/src/serde/de.rs:60]: This probably wouldn't work since we need the type name in order to deserialize the actual value (plus this makes it use struct format which might not be desired)

File: crates/bevy_reflect/src/serde/de.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/serde/de.rs:33]: I'm in favor of renaming to `Untyped`; it's clearer and indicates that there is at least one other option.

Can `TypedReflectDeserializer` deserialize to concrete types, since it has type info up front? If so, I think it would become the default deserializer, since that is what I and I think anyone else would expect a deserializer to do.

[the remark this answers is not in the document]
Comment (MrGVSV) [reply crates/bevy_reflect/src/serde/de.rs]: Went ahead with the rename!

**Comment by Metadorius** [comment]
My suggestion on making the serialization syntax even more human-processable: **omit `type` and `value` tags and express them as key-value pairs**.

### Example

Instead of
```json
{
  "type": "my_game::inventory::Inventory",
  "value": {
    "id": "inv001",
    "max_storage": 10,
    "items": [
      (
        id: "Pickaxe"
      )
    ],
  },
},
```
we would write
```json
"my_game::inventory::Inventory": {
  "id": "inv001",
  "max_storage": 10,
  "items": [
    (
      id: "Pickaxe"
    )
  ],
},
```

### Why?

TL;DR it is just a **less error-prone and more intuitive representation** without artificial to JSON / RON / YAML rules, instead these rules are expressed naturally through JSON / RON / YAML syntax.
1. With old implementation you needed to introduce artificial *"you have to write type before value"* rule. Ordering of key-values is unnatural for such data representations. With my suggestion you simply can't write value before key as there's no such syntax.
2. Same deal with *"you can't add multiple components of the same type"* rule. Because ECS (Bevy's in particular) don't allow for multiple components, the data structure that represents an ECS entity's components is a *component set* (or a type-value dictionary/map which is almost the same) and absolutely *not an array*. Current format uses arrays, my suggestion uses dictionary/set representation as you simply can't define the same key (type) twice on one object. I don't think using arrays to represent sets is a good idea.

### Alternatives considered

#4153 is a good pitch too and aims to solve the same issue, except the way that what @AronDerenyi proposed only solves one of the two inconsistencies. For RON it looks nicer, but it still allows for multiple components on a syntax level, which is semantically invalid and has to be limited, whereas what I propose doesn't allow for that. Also I don't think you really can express the same syntax with JSON, YAML etc. meaning that it is less portable.

**Comment by inodentry** [comment]
I really like this last suggestion here, by @Metadorius , for the format. :) Agreed with the reasoning. Seems like this is perhaps the best proposal we have so far.

**Comment by MrGVSV** [comment]
I updated the code and the PR description to follow @Metadorius' suggestion! It now uses `{type_name: data}` in place of `{type: type_name, value: data}`.

File: assets/scenes/load_scene_example.scn.ron [hunk 5]
```diff
@@ -4,41 +4,23 @@
...
-          "rotation": {
-            "type": "glam::quat::Quat",
-            "value": (0.0, 0.0, 0.0, 1.0),
-          },
-          "scale": {
-            "type": "glam::vec3::Vec3",
-            "value": (1.0, 1.0, 1.0),
-          },
+        "value": {
+          "translation": (0.0, 0.0, 0.0),
```
Comment (sarkahn) [inline assets/scenes/load_scene_example.scn.ron:8]:
```suggestion
          "translation": {
            "x" : 0.0, 
            "y": 0.0, 
            "z": 0.0
          },
```

File: assets/scenes/load_scene_example.scn.ron [hunk 6]
```diff
@@ -4,41 +4,23 @@
...
+          "rotation": (0.0, 0.0, 0.0, 1.0),
+          "scale": (1.0, 1.0, 1.0),
```
Comment (sarkahn) [inline assets/scenes/load_scene_example.scn.ron:10]:
```suggestion
          "scale": {
            "x": 1.0, 
            "y": 1.0, 
            "z": 1.0
          },
```

**Review by sarkahn — CHANGES_REQUESTED** [review]
Example scene needs to be updated to the new format. 

Really looking forward to this getting merged!

File: crates/bevy_reflect/src/serde/de.rs (hunk 1 above)
Comment (MrGVSV) [reply crates/bevy_reflect/src/serde/de.rs:209]: This has now been addressed in a separate PR, #5723

File: crates/bevy_reflect/src/serde/de.rs [hunk 7]
```diff
@@ -12,375 +16,476 @@ pub trait DeserializeValue {
...
-                    return Ok(value);
-                }
-                _ => return Err(de::Error::unknown_field(key.as_str(), &[])),
-            }
-        }
-
-        Err(de::Error::custom("Maps in this location must have the \'type\' field and one of the following fields: \'map\', \'seq\', \'value\'"))
-    }
+        let type_name = map
+            .next_key::<&str>()?
```
Comment (Suficio) [inline crates/bevy_reflect/src/serde/de.rs:141]: This change proves problematic, as `&str` will panic when deserializing `String`.

Suggest to deserialize `String` directly instead as this will make the deserializer more robust, and is what the previous deserializer used.

This is particularly painful when deserializing from foreign `Deserializer` objects.

File: crates/bevy_reflect/src/serde/de.rs [hunk 8]
```diff
@@ -390,221 +495,449 @@ impl<'a, 'de> Visitor<'de> for MapVisitor<'a> {
...
         V: MapAccess<'de>,
     {
-        let mut dynamic_struct = DynamicStruct::default();
-        while let Some(key) = map.next_key::<String>()? {
-            let value = map.next_value_seed(ReflectDeserializer {
-                registry: self.registry,
-            })?;
-            dynamic_struct.insert_boxed(&key, value);
+        let variant_name = map
+            .next_key::<&str>()?
```
Comment (Suficio) [inline crates/bevy_reflect/src/serde/de.rs:515]: I haven't encountered the same problem as in my previous comment, however, this is also a potential point of failure.

**Comment by alice-i-cecile** [comment]
Closed in favor of #5723 per https://github.com/bevyengine/bevy/pull/5723#discussion_r960251187 🎉
